packages/workspace/src/core/hasher/hasher.ts

```diff
@@ -148,12 +162,149 @@ export class Hasher {
+      ),
+      record: 'tsconfig.base.json[compilerOptions][paths]',
+    };
+
+    return [hashedTsConfig, hashedTsConfigPaths];
+  }
+
+  private workspaceJsonDepsHash(
+    sortedTransitiveDeps: Array<[string, ProjectGraphNode]>,
+    workspaceJson: any,
+    workspaceJsonName: string
+  ): {
+    hash: string;
+    record: string;
+  } {
+    const hashedWorkspaceJsonRecords: string = this.hashing.hashArray(
+      sortedTransitiveDeps.map(([name]) =>
+        this.hashing.hashObject(sortObject(workspaceJson.projects[name]))
+      )
+    );
+
+    return {
+      hash: hashedWorkspaceJsonRecords,
+      record: workspaceJsonName,
+    };
+  }
+
+  /**
+   * Will selectively combine all of the transitive deps of the current
+   * project based on the project graph
+   */
+  async transitiveClosureDepsHash(): Promise<ImplicitHashResult> {
+    const nodes = Object.entries(this.projectGraph.nodes);
+
+    const sortedTransitiveDeps: Array<[
+      string,
+      ProjectGraphNode
+    ]> = nodes
+      .filter(([name, records]) => !name.startsWith('npm:'))
+      .sort(([name1, records1], [name2, records2]) =>
+        name1.localeCompare(name2)
+      );
+
+    const workspaceJsonName = workspaceFileName();
+    const workspaceJson = readWorkspaceJson();
+
+    const tsConfigTransitiveHashes = this.tsconfigDepsHash(
+      sortedTransitiveDeps,
+      workspaceJson
+    );
+
+    const workspaceJsonTransitiveHashes = this.workspaceJsonDepsHash(
+      sortedTransitiveDeps,
+      workspaceJson,
+      workspaceJsonName
+    );
+
+    /**
+     * 1: Filter out npm dependencies, they will be accounted for when calculating hash key for package.json
+     * 2: Sort by the name of the project so that reorganizing nx.json does not cause cache misses
+     * 3. Hash each nx.json project independently
+     */
+    const hashedNxJsonRecords: Array<{
+      hash: string;
+      record: string;
+    }> = sortedTransitiveDeps.map(([name]) => {
+      const nxJsonRecord = this.nxJson.projects[name];
+
+      return {
+        hash: this.hashing.hashObject(nxJsonRecord),
+        record: name,
+      };
+    });
+
+    const combinedHashRecords: Array<{
+      hash: string;
+      record: string;
+    }> = [
+      ...hashedNxJsonRecords,
+      ...tsConfigTransitiveHashes,
+      workspaceJsonTransitiveHashes,
+    ];
+
+    return Promise.resolve().then(async () => {
+      return {
+        value: this.hashing.hashArray(
+          combinedHashRecords.map((hashed) => hashed.hash)
+        ),
+        sources: combinedHashRecords.reduce(
+          (prev, cur) => ({
+            ...prev,
+            [cur.record]: cur.hash,
+          }),
+          {}
+        ),
+      };
+    });
+  }
+
   private async implicitDepsHash(): Promise<ImplicitHashResult> {
     if (this.implicitDependencies) return this.implicitDependencies;
 
     const fileNames = [
       ...Object.keys(this.nxJson.implicitDependencies || {}),
-      ...rootWorkspaceFileNames(),
```

Comment:
Changed this to just package.json since `tsconfig.base.json` `nx.json` and `workspace.json` are handled in the other function.